### PR TITLE
Project 3: Second Milestone for TIC TAC TOE game.

### PR DESCRIPTION
**Pull request by FdI96**
In this Milestone I added the logic for the game in the following file:
-game_logic.rb

Adapted the user interface file with the logic game's file
-main.rb

Changed the README.md file
-README.md

### CODE REVIEW

**Review by mahmoud717 — CHANGES_REQUESTED** [review]
### Changes required ♻️ 

**Hello @FdI96**

Great job with the project so far it looks amazing, but there are still some changes to be made:
 
---
### Requested 🔴 

- [x] you should have each lib class in its own file.
- [ ] the player class shouldn't be inheriting the board class. if you want to access the methods and attributes of one of them you have to import the other in the file.
 that's what oop is all about, separating different components from each other so that it's easier to read and debug.

- [ ] there should be no puts or print outside of the main file.
- [x] you must use require_relative instead of require if you want to import files. require is for gems only.
- [ ] when it's a draw, please make sure to print it.

### Optional 🟢 

- [ ] It would be really nice if you added a function to ask the players if they want to play again after a game is finished. 

---

please make these changes and request another code review.

if you have any questions don't hesitate to contact me on slack **MahmoudMohammad**

**Review by mahmoud717 — CHANGES_REQUESTED** [review]
### Changes required ♻️ 

**Hello @FdI96**
 
you haven't fixed all of the requested issues, please make sure to fix the issues before asking for another code review.
---
### Requested 🔴 

- [ ] the player class shouldn't be inheriting the board class. if you want to access the methods and attributes of one of them you have to import the other in the file.
 that's what oop is all about, separating different components from each other so that it's easier to read and debug.

- [ ] there should be no puts or print outside of the main file.

- [ ] when it's a draw, please make sure to print it.

- [ ] the game should quit if a player didn't enter P to start. it should say unknown command please enter a valid command or something like this.


---

please make these changes and request another code review.

if you have any questions don't hesitate to contact me on slack **MahmoudMohammad**

**Review by mke2111 — CHANGES_REQUESTED** [review]
# **STATUS: CHANGES REQUIRED 🔴 ⚠️** 

## Hey @FdI96, 👋🏽 
 **Thank you for the good job you got going on. The game runs almost just fine. Just a few changes to make and we'll be right there.**

## REQUIRED CHANGES ⚠️ 

- Kindly implement **all** changes as requested by the previous reviewer. 

- [ ] Ensure that there are no `puts`, `gets.chomp` or `print` outside the `main.rb` file. Printing and getting user input can only be done from the `main.rb` file.

You need to transfer all methods/code that have `print`, `puts`, and `gets.chomp` into the the main file.

**Kindly make all the requested changes and submit another CRR. 👍🏽** 

## Cheers!! 🍹 & Happy Coding.

# Reach me on slack for a chitchat 🆘  **Roy Mukuye**

File: lib/board.rb [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  end
+
+  def turn(turn, player_one, player_two)
+    turn ? player_one : player_two
+  end
+
+  def updated_board
+    puts "#{board_array[0]} | #{board_array[1]} | #{board_array[2]}"
+    puts '-- --- --'
+    puts "#{board_array[3]} | #{board_array[4]} | #{board_array[5]}"
```
Comment (mke2111) [inline lib/board.rb]: Kindly transfer the printing logic to the main file

File: lib/board.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  def modify_board(number, turn)
+    board_array[number - 1] = turn ? 'X' : 'O'
+    combination_array.map! { |elem| elem.map! { |sub_elem| inside_array(sub_elem, number) } }
+  end
+
+  def win_or_draw(accu, player)
+    win = false
+    combination_array.each { |elem| elem.all?('X') || elem.all?('O') ? win = true : false }
+    if win
+      puts "Player #{player.name} WINS"
```
Comment (mke2111) [inline lib/board.rb]: Kindly transfer the printing logic to the main file

File: lib/player.rb [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+  def player_turn(board, turn)
+    band = false
+    band = select_slot(band, turn, board) until band
+  end
+
+  def select_slot(band, turn, board)
+    puts 'Please select one empty space'
+    board.updated_board
+    number = gets.chomp.to_i
```
Comment (mke2111) [inline lib/player.rb]: Kindly transfer the user input logic to the main file

**Review by jubaan — CHANGES_REQUESTED** [review]
[![Microverse][microverse-shield]][microverse-url]
<div>
<h1 align="center"> ⚠️ 🛠️ ⚠️ 🛠️ Changes Required 🛠️ ⚠️ 🛠️ ⚠️ </h1>
<br />
<p><b>"Fail fast, fail often"</b> and don't let this discourage you, because <b>only by failing do we learn to succeed</b>.</p>
<details open="open">
  <summary>Review Summary</summary>
  <ol>
    <li><a href="#congrats"> 💯 Awesome work 👏🏼 👏🏼 </a></li>
    <li><a href="#mandatory"> ⚠️ Things to Improve </a> 🚨 🚨 </li>
    <li><a href="#optional"> 💡Nice to have and other comments ✍🏼 ✍🏼 </a></li>
    <li><a href="#contact"> 🧔🏼 About your Reviewer </a> (you don't know it... but we are just automated bots 🤖) </li>
    <li><a href="#help"> 🆘 <b> 😭 Feeling stuck? or Need help</b>? 🙏🏼 🙏🏼 </a></li>
  </ol>
</details>
</div>

<h3 id="congrats"> 1. :100: Awesome work</h3>
<details close="close">
<summary> [click to open] Big or little achievements, but achievements at the end. <b>Feel proud that you're learning and improving everyday.</b></summary>


- [ ] You're almost done! 


</details>

<hr>
<h3 id="mandatory"> 2. ⚠️ Things to Improve (or the To-Do list to get the project approved) </h3> 
<details close="close">
<summary> [click to open] Working on these observations will help you advance and get closer to the goal 🎯, postpone them or do them half-ways and they'll haunt you 👻 👻... FOREVER!! (It will feel like that, trust me). </summary>


- [ ] All logic should be in the library directory. Move `updated_board`, `win_or_draw`. `select_slot` to their parent class or module.
- [ ] Only in the main.rb is allowed to use `prints`, and `puts`.
- [ ] Is not just moving the methods, but to take out the prints and puts from the logic and when application makes the call in the fronend, then it prints or outputs the returned value of the method.


</details>
<p><b>NOTE</b>This list is or could be divided in what is show in this comment and inline comments posted below.<p>
<hr>

<h3 id="optional"> 3. 💡 Nice to have and other comments </h3>
<details close="close">
<summary> [click to open] </h3></summary>


- [ ] Nothing more to add.


</details>

<h3 id="contact"> 4. 🧔🏼 About your Reviewer </h3>
<details close="close">
<summary> [click to open] Say "hello world!" on Slack: @jubaan</summary>
<br />
<div align="center">
  <a src="https://www.github.com/jubaan">
    Julio Añoveros
  </a>
  <p>
    Julio Añoveros | @jubaan</br></br>
    <b>Full-stack Web Developer</b> and</br>
    🤖 Code Reviewer at <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a>:rocket:
  </p>
</div>
</details>

<h3 id="help"> 5. 🆘 <b> 😭 Feeling stuck? or Need help</b>? 🙏🏼 🙏🏼</h3>
<details close="close">
  <summary> [click to open] <b>Feel stuck?</b> </summary>
  <br>
  <p> First breath... just do it. Take 15-20 min away from the computer, stretch, meditate, do some exercise, drink water, or do whatever is necessary to distract your mind from the obstacle.</p>
  <p><b>Sometimes you just need to get away to see things with a new perspective.</b></p>
</details>

<details close="close">
  <summary> [click to open] <b>Need help?</b> </summary>
  <br>
  <p> Send me a DM in <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a> Students Slack, you'll find me as @jubaan. If I'm not available there are other Code Reviewers you can contact. They'll help as well.</p>
  <p>Remember that <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a> has other channels where you can get an answer to your questions:</p>
  <ol>
    <li>The <b>students channel 🧑‍🎓 </b> in slack,</li>
    <li>your <b>teammates 👥 </b>,</li>
    <li>your <b>mentor 👨🏼 👩🏼 </b>, and</li>
    <li><b>Student Success 🏫 </b>.</li>
  </ol>
  <p><b> DON'T hesitate to ask! </b></p>
</details>
<p>If all of the above doesn't help... get a <a href="https://rubberduckdebugging.com/">rubber duck</a> ... it would be great to own one 😄. </p>

<h2 align="center"> Keep up the great work :muscle: and Happy Coding :smiley: :computer:!! </h1>

[microverse-shield]: https://img.shields.io/badge/-microverse-%236F23FF?style=for-the-badge
[microverse-url]: https://www.microverse.org/?grsf=kfm2ej
[bootstrap-shield]: https://img.shields.io/badge/Bootstrap-white?style=for-the-badge&logo=bootstrap
[bootstrap-url]: https://www.microverse.org/?grsf=kfm2ej

File: bin/main.rb [hunk 4]
```diff
@@ -8,108 +9,103 @@ def updated_board(board_array)
...
+    if win
+      puts "Player #{player.name} WINS"
+      false
+    elsif accu == 9
+      puts 'No more plays available. Its a DRAW'
+      false
+    else
+      true
+    end
+  end
```
Comment (jubaan) [inline bin/main.rb]: - [ ] This logic should be inside the `lib` directory. It also should not contain `puts` nor `prints`.

File: bin/main.rb [hunk 5]
```diff
@@ -8,108 +9,103 @@ def updated_board(board_array)
...
+  def select_slot(band, turn, board)
+    puts 'Please select one empty space'
+    number = gets.chomp.to_i
+    if (1..9).include?(number)
+      if board.board_array[number - 1] == 'X' || board.board_array[number - 1] == 'O'
+        puts 'Slot occupied'
+      else
+        board.modify_board(number, turn)
+        band = true
+      end
```
Comment (jubaan) [inline bin/main.rb]: - [ ] This logic should be inside the `lib` directory. It also should not contain `puts` nor `prints`.

File: bin/main.rb [hunk 6]
```diff
@@ -1,4 +1,5 @@
 #!/usr/bin/env ruby
+require_relative '../lib/board.rb'
 
 def updated_board(board_array)
```
Comment (jubaan) [inline bin/main.rb]: - [ ] This logic should be inside the `lib` directory. It also should not contain `puts` nor `prints`.

**Review by jubaan — CHANGES_REQUESTED** [review]
[![Microverse][microverse-shield]][microverse-url]
<div>
<h1 align="center"> ⚠️ 🛠️ ⚠️ 🛠️ Changes Required 🛠️ ⚠️ 🛠️ ⚠️ </h1>
<br />
<p><b>"Fail fast, fail often"</b> and don't let this discourage you, because <b>only by failing do we learn to succeed</b>.</p>
<details open="open">
  <summary>Review Summary</summary>
  <ol>
    <li><a href="#congrats"> 💯 Awesome work 👏🏼 👏🏼 </a></li>
    <li><a href="#mandatory"> ⚠️ Things to Improve </a> 🚨 🚨 </li>
    <li><a href="#optional"> 💡Nice to have and other comments ✍🏼 ✍🏼 </a></li>
    <li><a href="#contact"> 🧔🏼 About your Reviewer </a> (you don't know it... but we are just automated bots 🤖) </li>
    <li><a href="#help"> 🆘 <b> 😭 Feeling stuck? or Need help</b>? 🙏🏼 🙏🏼 </a></li>
  </ol>
</details>
</div>

<h3 id="congrats"> 1. :100: Awesome work</h3>
<details close="close">
<summary> [click to open] Big or little achievements, but achievements at the end. <b>Feel proud that you're learning and improving everyday.</b></summary>



</details>

<hr>
<h3 id="mandatory"> 2. ⚠️ Things to Improve (or the To-Do list to get the project approved) </h3> 
<details close="close">
<summary> [click to open] Working on these observations will help you advance and get closer to the goal 🎯, postpone them or do them half-ways and they'll haunt you 👻 👻... FOREVER!! (It will feel like that, trust me). </summary>


- [ ] The same issues still persist please review your `main.rb` file.
> * [ ] All logic should be in the library directory. Move `updated_board`, `win_or_draw`. `select_slot` to their parent class or module.
> * [ ]  Only in the main.rb is allowed to use `prints`, and `puts`.
> * [ ]  Is not just moving the methods, but to take out the prints and puts from the logic and when application makes the call in the fronend, then it prints or outputs the returned value of the method.




</details>
<p><b>NOTE</b>This list is or could be divided in what is show in this comment and inline comments posted below.<p>
<hr>

<h3 id="optional"> 3. 💡 Nice to have and other comments </h3>
<details close="close">
<summary> [click to open] </h3></summary>


- [ ] Nothing more to add.


</details>

<h3 id="contact"> 4. 🧔🏼 About your Reviewer </h3>
<details close="close">
<summary> [click to open] Say "hello world!" on Slack: @jubaan</summary>
<br />
<div align="center">
  <a src="https://www.github.com/jubaan">
    Julio Añoveros
  </a>
  <p>
    Julio Añoveros | @jubaan</br></br>
    <b>Full-stack Web Developer</b> and</br>
    🤖 Code Reviewer at <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a>:rocket:
  </p>
</div>
</details>

<h3 id="help"> 5. 🆘 <b> 😭 Feeling stuck? or Need help</b>? 🙏🏼 🙏🏼</h3>
<details close="close">
  <summary> [click to open] <b>Feel stuck?</b> </summary>
  <br>
  <p> First breath... just do it. Take 15-20 min away from the computer, stretch, meditate, do some exercise, drink water, or do whatever is necessary to distract your mind from the obstacle.</p>
  <p><b>Sometimes you just need to get away to see things with a new perspective.</b></p>
</details>

<details close="close">
  <summary> [click to open] <b>Need help?</b> </summary>
  <br>
  <p> Send me a DM in <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a> Students Slack, you'll find me as @jubaan. If I'm not available there are other Code Reviewers you can contact. They'll help as well.</p>
  <p>Remember that <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a> has other channels where you can get an answer to your questions:</p>
  <ol>
    <li>The <b>students channel 🧑‍🎓 </b> in slack,</li>
    <li>your <b>teammates 👥 </b>,</li>
    <li>your <b>mentor 👨🏼 👩🏼 </b>, and</li>
    <li><b>Student Success 🏫 </b>.</li>
  </ol>
  <p><b> DON'T hesitate to ask! </b></p>
</details>
<p>If all of the above doesn't help... get a <a href="https://rubberduckdebugging.com/">rubber duck</a> ... it would be great to own one 😄. </p>

<h2 align="center"> Keep up the great work :muscle: and Happy Coding :smiley: :computer:!! </h1>

[microverse-shield]: https://img.shields.io/badge/-microverse-%236F23FF?style=for-the-badge
[microverse-url]: https://www.microverse.org/?grsf=kfm2ej
[bootstrap-shield]: https://img.shields.io/badge/Bootstrap-white?style=for-the-badge&logo=bootstrap
[bootstrap-url]: https://www.microverse.org/?grsf=kfm2ej

**Comment by FdI96** [comment]
Hi there ! I’ve asked you some things on slack.

**Review by descholar-ceo — CHANGES_REQUESTED** [review]
# Status: Changes Required :vertical_traffic_light: :warning: 
## Hello @FdI96  :handshake: :handshake: :handshake: 

:+1: This is good to make it this far, your project is really nice, congratulations on the good work you have done so far.
:+1: I liked that you have tried to implemented the requested changes from the previous review, keep it up, and your project is really almost done.
#### :warning: However we need to make few changes to mark this project approved, don't feel discouraged because of being asked some changes more and more, that's beacuse this is a learning process, and we are here to help you to become an out-standing developer. See the requested change below :point_down: :point_down: :point_down: 
### Note: All of requested changes which are marked as [Required] are mandatory to be implemented in order to get this project approved, others which are marked as [Optional] are nice to have in your project to make it outstanding

# [Required Changes] :red_circle: 
- [x] Move thev`updated_board` method to lib folder
- [x] When there is a win, please display the updated board containing the win move
- [x] validate the names of players, don't allow empty names
## Optional :yellow_circle: 
- [ ] It would be very nice if the game doesn't get terminated after a win/draw, it rather asks players whether they want to play another round or not,
___
_If anything seems to be unclear to you, reachout and ask anything by commenting on this Pull Request and mention me as [descholar-ceo](https://github.com/descholar-ceo), I will receive a notification and will respond._

[GitHub](https://github.com/descholar-ceo) | [LinkedIn](https://linkedin.com/in/mugirase-emmanuel) | [Twitter](https://twitter.com/descholar3)

_Reviewed by [Mugirase Emmanuel (descholar)](https://github.com/descholar-ceo)_

File: lib/tools.rb [hunk 7]
```diff
@@ -0,0 +1,7 @@
+def updated_board(board_array)
+  puts "#{board_array[0]} | #{board_array[1]} | #{board_array[2]}"
+  puts '-- --- --'
+  puts "#{board_array[3]} | #{board_array[4]} | #{board_array[5]}"
+  puts '-- --- --'
+  puts "#{board_array[6]} | #{board_array[7]} | #{board_array[8]}"
+end
```
Comment (rindrajosia) [inline lib/tools.rb]: No `puts/print` and `gets.chomp` outside `main.rb` file. Please, move this in your `main.rb` file.

**Review by rindrajosia — CHANGES_REQUESTED** [review]
**Required Changes**
Hi, I'm Josia your reviewer for this PR
Good job improving your project! There are still some issues to revise before approving this project but hang in there and read the comments below with care. You can do it! :muscle: 

- [x] [REQUIRED] Your game should not `exit` or `raise an error` if the user enters an empty name, it should ask him `to enter his name again`.


- [x]  [REQUIRED] Your game should `not raise an error` if the user enters a letter different from 'S' when you ask him to start the game. It should just stop with a message like 'Goodbye'.


- [x]  [REQUIRED] No `puts/print` and `gets.chomp` outside `main.rb` file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

_This project is almost done!_
Happy coding
Rindra Josia

**Review by andresporras3423 — CHANGES_REQUESTED** [review]
### Status: Changes Required :recycle:
**Hello Federico Lopez Cechini** 
Good job with the improvements for this project :muscle:  :muscle:
We still need to make some improvements before you can go ahead:

- [ ] Fix rubocop offenses
- [ ] The instruction says "Press 'S' to start" but I guess you actually mean "Type 'S' to start" (the same for "Press 'P' to start")
- [ ] Validate player1 and player2 have different names
- [ ] In Board class, combination_array is never directly read outside the class so it doesn't need the attr_reader.
- [ ] In the main.rb file, the "if" on line 47 has too many responsibilities, try to split its logic into different functions.
- [ ] In the main.rb file, try to organize your code in function with descriptive names.
 
Please request another review once you are ready! :+1: 
 
Happy coding :relaxed:
 
Oscar Andrés Russi Porras

**Comment by FdI96** [comment]
Hey Oscar! I've talked to you on Slack.

**Review by kevcap — APPROVED** [review]
# Great job! Project approved! 💪

Any questions feel free to reach me through slack.

**Pull request merged**